### PR TITLE
Add some network-related definitions

### PR DESCRIPTION
**Pull request by zhhyu7**
## Summary
We will integrate network management tools in future projects.  In the process of porting the tools, we encounter some situations where the structure is not defined, or the returned data types do not match the expectations.  Refer to the common implementation of other systems and add relevant definitions.
Other patches will be submitted later, replacing some similar structures used in NuttX with common structures. will not replace structures that are specifically aligned with 1 or 2 bytes.

### CODE REVIEW

File: include/net/if_arp.h [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NET_IF_ARP_H_
+#define __INCLUDE_NET_IF_ARP_H_
```
Comment (pkarashchenko) [inline include/net/if_arp.h:22]: please fix the style

File: include/net/ethernet.h [hunk 2]
```diff
@@ -33,6 +33,10 @@
 
 #define ETHER_ADDR_LEN  6
 
+#define ETHERTYPE_ARP   0x0806 /* Address resolution protocol */
+#define ETHERTYPE_IP    0x0800 /* IP */
+#define ETHERTYPE_IPV6  0x86dd /* IP protocol version 6 */
```
Comment (pkarashchenko) [inline include/net/ethernet.h:38]: why defines from `include/nuttx/net/ethernet.h` can't be used?

File: include/net/if_arp.h [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+ */
+
+struct arphdr
+{
+  unsigned short int ar_hrd;          /* Format of hardware address.  */
+  unsigned short int ar_pro;          /* Format of protocol address.  */
+  unsigned char ar_hln;               /* Length of hardware address.  */
+  unsigned char ar_pln;               /* Length of protocol address.  */
+  unsigned short int ar_op;           /* ARP opcode (command).  */
+};
```
Comment (pkarashchenko) [inline include/net/if_arp.h:133]: Why `struct arp_hdr_s` can't be used?

File: include/netinet/icmp6.h [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+  ((((filterp)->icmp6_filt[(type) >> 5]) & (1 << ((type) & 31))) != 0)
+
+#define ICMP6_FILTER_SETPASS(type, filterp) \
+  ((((filterp)->icmp6_filt[(type) >> 5]) &= ~(1 << ((type) & 31))))
+
+#define ICMP6_FILTER_SETBLOCK(type, filterp) \
+  ((((filterp)->icmp6_filt[(type) >> 5]) |=  (1 << ((type) & 31))))
+
+#define ICMP6_FILTER_SETPASSALL(filterp) \
+  memset (filterp, 0, sizeof (struct icmp6_filter));
```
Comment (pkarashchenko) [inline include/netinet/icmp6.h:110]:
```suggestion
  memset(filterp, 0, sizeof (struct icmp6_filter));
```

File: include/netinet/icmp6.h [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+
+#define ICMP6_FILTER_SETBLOCKALL(filterp) \
+  memset (filterp, 0xFF, sizeof (struct icmp6_filter));
```
Comment (pkarashchenko) [inline include/netinet/icmp6.h:113]:
```suggestion
  memset(filterp, 0xFF, sizeof (struct icmp6_filter));
```

File: include/netinet/icmp6.h [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+
+#define ND_ROUTER_SOLICIT           133
+#define ND_ROUTER_ADVERT            134
+#define ND_NEIGHBOR_SOLICIT         135
+#define ND_NEIGHBOR_ADVERT          136
+#define ND_REDIRECT                 137
+
+struct nd_router_solicit      /* router solicitation */
+{
+  struct icmp6_hdr  nd_rs_hdr;
```
Comment (pkarashchenko) [inline include/netinet/icmp6.h:123]:
```suggestion
  struct icmp6_hdr nd_rs_hdr;
```

File: include/netinet/icmp6.h [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+#define nd_rs_type               nd_rs_hdr.icmp6_type
+#define nd_rs_code               nd_rs_hdr.icmp6_code
+#define nd_rs_cksum              nd_rs_hdr.icmp6_cksum
+#define nd_rs_reserved           nd_rs_hdr.icmp6_data32[0]
+
+struct nd_router_advert       /* router advertisement */
+{
+  struct icmp6_hdr  nd_ra_hdr;
+  uint32_t   nd_ra_reachable;   /* reachable time */
+  uint32_t   nd_ra_retransmit;  /* retransmit timer */
```
Comment (pkarashchenko) [inline include/netinet/icmp6.h:137]:
```suggestion
  struct icmp6_hdr nd_ra_hdr;
  uint32_t nd_ra_reachable;   /* reachable time */
  uint32_t nd_ra_retransmit;  /* retransmit timer */
```

File: include/netinet/icmp6.h [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+#define nd_ra_flags_reserved     nd_ra_hdr.icmp6_data8[1]
+#define ND_RA_FLAG_MANAGED       0x80
+#define ND_RA_FLAG_OTHER         0x40
+#define ND_RA_FLAG_HOME_AGENT    0x20
+#define nd_ra_router_lifetime    nd_ra_hdr.icmp6_data16[1]
+
+struct nd_neighbor_solicit    /* neighbor solicitation */
+{
+  struct icmp6_hdr  nd_ns_hdr;
+  struct in6_addr   nd_ns_target; /* target address */
```
Comment (pkarashchenko) [inline include/netinet/icmp6.h:155]:
```suggestion
  struct icmp6_hdr nd_ns_hdr;
  struct in6_addr nd_ns_target; /* target address */
```

File: include/netinet/icmp6.h [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+#define nd_ns_type               nd_ns_hdr.icmp6_type
+#define nd_ns_code               nd_ns_hdr.icmp6_code
+#define nd_ns_cksum              nd_ns_hdr.icmp6_cksum
+#define nd_ns_reserved           nd_ns_hdr.icmp6_data32[0]
+
+struct nd_neighbor_advert     /* neighbor advertisement */
+{
+  struct icmp6_hdr  nd_na_hdr;
+  struct in6_addr   nd_na_target; /* target address */
```
Comment (pkarashchenko) [inline include/netinet/icmp6.h:168]:
```suggestion
  struct icmp6_hdr nd_na_hdr;
  struct in6_addr nd_na_target; /* target address */
```

File: include/netinet/ip.h [hunk 10]
```diff
@@ -32,6 +32,33 @@
...
+#define IPDEFTTL  64    /* default ttl, from RFC 1340 */
+
+struct iphdr
+{
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+  unsigned int ihl:4;
+  unsigned int version:4;
+#elif __BYTE_ORDER == __BIG_ENDIAN
+  unsigned int version:4;
+  unsigned int ihl:4;
```
Comment (pkarashchenko) [inline include/netinet/ip.h:45]: I strongly do not recommend using bit fields for network processing

File: include/net/if_arp.h (hunk 1 above)
Comment (zhhyu7) [reply include/net/if_arp.h]: done

File: include/net/ethernet.h (hunk 2 above)
Comment (zhhyu7) [reply include/net/ethernet.h:38]: glibc and musl both define ETHERTYPE_*, so many applications use this definition, and if we compile directly in NuttX's environment, we will get a compile error.

File: include/net/if_arp.h (hunk 3 above)
Comment (zhhyu7) [reply include/net/if_arp.h:133]: glibc and musl both define arphdr, and the structure and names of the members are different from what NuttX already has. so many applications use this definition, and if we compile directly in NuttX's environment, we will get a compile error.

File: include/netinet/icmp6.h (hunk 4 above)
Comment (zhhyu7) [reply include/netinet/icmp6.h]: done

File: include/netinet/icmp6.h (hunk 5 above)
Comment (zhhyu7) [reply include/netinet/icmp6.h]: done

File: include/netinet/icmp6.h (hunk 6 above)
Comment (zhhyu7) [reply include/netinet/icmp6.h]: done

File: include/netinet/icmp6.h (hunk 7 above)
Comment (zhhyu7) [reply include/netinet/icmp6.h]: done

File: include/netinet/icmp6.h (hunk 8 above)
Comment (zhhyu7) [reply include/netinet/icmp6.h]: done

File: include/netinet/icmp6.h (hunk 9 above)
Comment (zhhyu7) [reply include/netinet/icmp6.h]: done

File: include/netinet/ip.h (hunk 10 above)
Comment (zhhyu7) [reply include/netinet/ip.h:45]: glibc and musl both define this bit fields, so many applications use this definition, and if we compile those applications directly in NuttX's environment, we will get a compile error. this structure is only for tripartite library compatibility and is not used by NuttX itself

File: include/net/ethernet.h (hunk 2 above)
Comment (hartmannathan) [reply include/net/ethernet.h:38]: Maybe the defines in `include/nuttx/net/ethernet.h` should be moved here, to be similar to glibc and musl in this regard? (FreeBSD has it in sys/net/ethernet.h -- I am not sure if that's the include path programs use, or if they use net/ethernet.h without the sys.)

Comment (hartmannathan) [reply include/net/ethernet.h:38]: Another idea: With Kconfig, developers could select: NuttX POSIX-only, Linux-style, and/or BSD-style. These would not be mutually exclusive; that is, you could select more than one. Those selection(s) would enable/disable things like glibc-style declarations. Two disadvantages of this idea: (1) more Kconfig/ifdef in the codebase, (2) how do we handle cases where Linux-style and BSD-style conflict with each other?

Comment (xiaoxiang781216) [reply include/net/ethernet.h:38]: If no conflict happens, what's benefit to add option?

**Pull request closed without merging**